# 作業 : 調整 ELMo 模型的不同訓練參數

# [作業目標]
- 調整 ELMo 模型的不同參數, 分別觀察 loss 數據並比較

# [作業重點]
- 調整 ELMo 模型的不同訓練參數, 分別觀察 loss 數據並比較

# [參數說明]
- UNITS : LSTM 特徵維度，即每一筆訓練輸入單字的個數
- N_LAYERS : LSTM 堆疊的層數
- BATCH_SIZE : 訓練批次大小，即幾筆資料合併做一次倒傳遞
- LEARNING_RATE : 學習速率，影響收斂的快慢，須配合 BATCH_SIZE 調整

# 程式說明
- 程式採用 tensorflow2 / keras 寫作, 執行前請先安裝 tensorflow 2.0
- 本程式執行時, 請將 utils.py 與執行檔放置於同一目錄下
- 程式來源 : 莫煩Python-ELMo:一詞多義 https://mofanpy.com/tutorials/machine-learning/nlp/elmo/

In [1]:
# 連接個人資料 讀取 ＰＴＴ 訓練資料和儲存模型
#先連接自己的GOOGLE DRIVE 為了要儲存資料和訓練模型
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import os

# Current directory
print(os.getcwd())

# change directory
os.chdir('/content/drive/MyDrive/python_training/NLP100Days-part2/D28_ELMO/')
print(os.getcwd())

/content
/content/drive/MyDrive/python_training/NLP100Days-part2/D28_ELMO


In [3]:
# [Deep contextualized word representations](https://arxiv.org/pdf/1802.05365.pdf)
from tensorflow import keras
import tensorflow as tf
import utils    # this refers to utils.py in my [repo](https://github.com/MorvanZhou/NLP-Tutorials/)
import time
import os

class ELMo(keras.Model):
    def __init__(self, v_dim, emb_dim, units, n_layers, lr):
        super().__init__()
        self.n_layers = n_layers
        self.units = units

        # encoder
        self.word_embed = keras.layers.Embedding(
            input_dim=v_dim, output_dim=emb_dim,  # [n_vocab, emb_dim]
            embeddings_initializer=keras.initializers.RandomNormal(0., 0.001),
            mask_zero=True,
        )
        # forward lstm
        self.fs = [keras.layers.LSTM(units, return_sequences=True) for _ in range(n_layers)]
        self.f_logits = keras.layers.Dense(v_dim)
        # backward lstm
        self.bs = [keras.layers.LSTM(units, return_sequences=True, go_backwards=True) for _ in range(n_layers)]
        self.b_logits = keras.layers.Dense(v_dim)

        self.cross_entropy1 = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
        self.cross_entropy2 = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
        self.opt = keras.optimizers.Adam(lr)

    def call(self, seqs):
        embedded = self.word_embed(seqs)        # [n, step, dim]
        """
        0123    forward
        1234    forward predict
         1234   backward 
         0123   backward predict
        """
        mask = self.word_embed.compute_mask(seqs)
        fxs, bxs = [embedded[:, :-1]], [embedded[:, 1:]]
        for fl, bl in zip(self.fs, self.bs):
            fx = fl(
                fxs[-1], mask=mask[:, :-1], initial_state=fl.get_initial_state(fxs[-1])
            )           # [n, step-1, dim]
            bx = bl(
                bxs[-1], mask=mask[:, 1:], initial_state=bl.get_initial_state(bxs[-1])
            )  # [n, step-1, dim]
            fxs.append(fx)      # predict 1234
            bxs.append(tf.reverse(bx, axis=[1]))    # predict 0123
        return fxs, bxs

    def step(self, seqs):
        with tf.GradientTape() as tape:
            fxs, bxs = self.call(seqs)
            fo, bo = self.f_logits(fxs[-1]), self.b_logits(bxs[-1])
            loss = (self.cross_entropy1(seqs[:, 1:], fo) + self.cross_entropy2(seqs[:, :-1], bo))/2
        grads = tape.gradient(loss, self.trainable_variables)
        self.opt.apply_gradients(zip(grads, self.trainable_variables))
        return loss, (fo, bo)

    def get_emb(self, seqs):
        fxs, bxs = self.call(seqs)
        xs = [tf.concat((f[:, :-1, :], b[:, 1:, :]), axis=2).numpy() for f, b in zip(fxs, bxs)]
        for x in xs:
            print("layers shape=", x.shape)
        return xs


def train(model, data, step):
    t0 = time.time()
    for t in range(step):
        seqs = data.sample(BATCH_SIZE)
        loss, (fo, bo) = model.step(seqs)
        if t % 80 == 0:
            fp = fo[0].numpy().argmax(axis=1)
            bp = bo[0].numpy().argmax(axis=1)
            t1 = time.time()
            print(
                "\n\nstep: ", t,
                "| time: %.2f" % (t1 - t0),
                "| loss: %.3f" % loss.numpy(),
                "\n| tgt: ", " ".join([data.i2v[i] for i in seqs[0] if i != data.pad_id]),
                "\n| f_prd: ", " ".join([data.i2v[i] for i in fp if i != data.pad_id]),
                "\n| b_prd: ", " ".join([data.i2v[i] for i in bp if i != data.pad_id]),
                )
            t0 = t1
    os.makedirs("./visual/models/elmo", exist_ok=True)
    model.save_weights("./visual/models/elmo/model.ckpt")


def export_w2v(model, data):
    model.load_weights("./visual/models/elmo/model.ckpt")
    emb = model.get_emb(data.sample(4))
    print(emb)


if __name__ == "__main__":
    utils.set_soft_gpu(True)
    UNITS = 256
    N_LAYERS = 6  ##2
    BATCH_SIZE = 16
    LEARNING_RATE = 2e-3
    d = utils.MRPCSingle("./MRPC", rows=2000)
    print("num word: ", d.num_word)
    m = ELMo(d.num_word, emb_dim=UNITS, units=UNITS, n_layers=N_LAYERS, lr=LEARNING_RATE)
    train(m, d, 10000)
    export_w2v(m, d)

1 Physical GPUs, 1 Logical GPUs
num word:  12880


step:  0 | time: 6.94 | loss: 9.463 
| tgt:  <GO> <quote> the u.s. will find the killers and they will learn the meaning of american justice , <quote> bush told the crowd , which burst into applause . <SEP> 
| f_prd:  twenty-eight flaking childhood drawn drawn drawn drawn drawn drawn drawn drawn drawn drawn drawn keating keating keating keating 1960s 1960s 1960s 1960s 1960s 1960s 1960s 1960s then then then then 
| b_prd:  hospital hospital hospital hospital hospital hospital hospital hospital hospital hospital custody custody thursdays thursdays thursdays thursdays pharmaceuticals hebron hebron hebron hebron hebron hebron exacerbate allison kilometres kilometres kilometres kilometres


step:  80 | time: 12.96 | loss: 7.898 
| tgt:  <GO> the worst affected area was south kyeongsang province where at least <NUM> people drowned and roads were swept away in mud sides . <SEP> 
| f_prd:  the , , , , , , , , , , , , , , , , , , , , , , , , 
|

num word:  12880

step:  0 | time: 3.43 | loss: 9.463 

| tgt:  <GO> at <NUM> p.m. edt , grace 's center was near latitude <NUM> north , longitude <NUM> west or about <NUM> miles east-southeast of corpus christi . <SEP> 

| f_prd:  unix cancun qanbar offenses offenses they they mud cover tragedy tragedy bbc russian russian medicines medicines medicines medicines idea reversal reversal reversal prequel prequel morbidly campbell dramatic dramatic dramatic dramatic dramatic dramatic dramatic dramatic dramatic dramatic dramatic 

| b_prd:  competitor competitor competitor competitor competitor abolished selling selling selling parker parker ballot ballot ballot ten ten congresses eagle quarantined quarantined quarantined prevailed prevailed suggests doubleheader doubleheader doubleheader